File: scripts/audit-handlers.mjs
import fs from 'fs';
import path from 'path';

const root = path.dirname(new URL(import.meta.url).pathname.replace(/^\/([A-Za-z]:)/, '$1'));
const pages = ['index.html', 'stage.html', 'ship.html', 'reports.html', 'contacts.html', 'notifications.html'];

const jsFiles = fs.readdirSync(root).filter(f => f.endsWith('.js'));
const jsContent = jsFiles.map(f => fs.readFileSync(path.join(root, f), 'utf8')).join('\n');
const defined = new Set([...jsContent.matchAll(/window\.(\w+)\s*=\s*(?:async\s+)?function/g)].map(m => m[1]));

const handlerRe = /window\.(\w+)\s*\(/g;
const inlineRe = /on(?:click|input|change|submit)\s*=\s*["']([^"']+)["']/gi;

let issues = 0;

for (const page of pages) {
  const html = fs.readFileSync(path.join(root, page), 'utf8');
  const scripts = [...html.matchAll(/<script src="([^"]+\.js)/g)].map(m => m[1].split('?')[0]);
  const handlers = new Set();
  for (const m of html.matchAll(handlerRe)) handlers.add(m[1]);
  for (const m of html.matchAll(inlineRe)) {
    for (const wm of m[1].matchAll(/window\.(\w+)/g)) handlers.add(wm[1]);
  }

  const pageJs = scripts.map(s => {
    const p = path.join(root, s);
    return fs.existsSync(p) ? fs.readFileSync(p, 'utf8') : '';
  }).join('\n');
  const pageDefined = new Set([...pageJs.matchAll(/window\.(\w+)\s*=\s*(?:async\s+)?function/g)].map(m => m[1]));

  const missing = [...handlers].filter(h => !pageDefined.has(h) && !defined.has(h)).sort();
  console.log(`\n=== ${page} (${scripts.length} scripts, ${handlers.size} handlers) ===`);
  if (missing.length) {
    issues += missing.length;
    console.log('MISSING:', missing.join(', '));
  } else {
    console.log('OK — all inline handlers resolve to a loaded script');
  }
}

console.log(`\n${issues ? `Found ${issues} unresolved handler(s)` : 'All pages passed handler audit'}`);
process.exit(issues ? 1 : 0);

File: scripts/audit-handlers.py
import re
import sys
from pathlib import Path

root = Path(__file__).resolve().parent.parent
pages = ['index.html', 'stage.html', 'ship.html', 'reports.html', 'contacts.html', 'notifications.html']

js_files = list(root.glob('*.js'))
js_content = '\n'.join(f.read_text(encoding='utf-8') for f in js_files)
defined = set(re.findall(r'window\.(\w+)\s*=\s*(?:async\s+)?function', js_content))

handler_re = re.compile(r'window\.(\w+)\s*\(')
inline_re = re.compile(r'on(?:click|input|change|submit)\s*=\s*["\']([^"\']+)["\']', re.I)

issues = 0
for page in pages:
    html = (root / page).read_text(encoding='utf-8')
    scripts = [m.split('?')[0] for m in re.findall(r'<script src="([^"]+\.js)', html)]
    handlers = set(handler_re.findall(html))
    for m in inline_re.finditer(html):
        handlers.update(re.findall(r'window\.(\w+)', m.group(1)))

    page_js = ''
    for s in scripts:
        p = root / s
        if p.exists():
            page_js += p.read_text(encoding='utf-8') + '\n'
    page_defined = set(re.findall(r'window\.(\w+)\s*=\s*(?:async\s+)?function', page_js))

    missing = sorted(h for h in handlers if h not in page_defined and h not in defined)
    print(f'\n=== {page} ({len(scripts)} scripts, {len(handlers)} handlers) ===')
    if missing:
        issues += len(missing)
        print('MISSING:', ', '.join(missing))
    else:
        print('OK — all inline handlers resolve to a loaded script')

print(f"\n{'Found ' + str(issues) + ' unresolved handler(s)' if issues else 'All pages passed handler audit'}")
sys.exit(1 if issues else 0)
